File: app/src/main/java/org/nitri/opentopo/adapter/WayPointListAdapter.java
package org.nitri.opentopo.adapter;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import org.nitri.opentopo.R;
import org.nitri.opentopo.model.WayPointHeaderItem;
import org.nitri.opentopo.model.WayPointItem;
import org.nitri.opentopo.model.WayPointListItem;

import java.util.List;

public class WayPointListAdapter
        extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private final List<WayPointListItem> mItems;
    private final OnItemClickListener mListener;

    public WayPointListAdapter(List<WayPointListItem> items, OnItemClickListener listener) {
        mItems = items;
        mListener = listener;
    }

    @Override
    public int getItemViewType(int position) {
        return mItems.get(position).getListItemType();
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int type) {
        View view;
        switch (type) {
            case WayPointListItem.HEADER:
                view = LayoutInflater
                        .from(viewGroup.getContext())
                        .inflate(R.layout.way_point_header_item, viewGroup, false);
                return new ViewHolderHeader(view);
            case WayPointListItem.WAY_POINT:
                view = LayoutInflater
                        .from(viewGroup.getContext())
                        .inflate(R.layout.way_point_item, viewGroup, false);
                return new ViewHolderWayPoint(view);
        }
        //noinspection ConstantConditions
        return null;
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder viewHolder, int pos) {
        WayPointListItem item = mItems.get(pos);
        ((ViewHolder) viewHolder).bindType(item);
    }

    @Override
    public int getItemCount() {
        return mItems.size();
    }

    public abstract class ViewHolder extends RecyclerView.ViewHolder {
        ViewHolder(View itemView) {
            super(itemView);
        }

        public abstract void bindType(WayPointListItem item);
    }

    public class ViewHolderHeader extends ViewHolder {
        private final TextView textView;

        ViewHolderHeader(View itemView) {
            super(itemView);
            textView = itemView.findViewById(R.id.textView);
        }

        public void bindType(WayPointListItem item) {
            textView.setText(((WayPointHeaderItem) item).getHeader());
        }
    }

    public class ViewHolderWayPoint extends ViewHolder implements View.OnClickListener {
        private final TextView textView;

        ViewHolderWayPoint(View itemView) {
            super(itemView);
            textView = itemView.findViewById(R.id.textView);
            itemView.setOnClickListener(this);
        }

        public void bindType(WayPointListItem item) {
            textView.setText(((WayPointItem) item).getWayPoint().getName());
            int itemIndex = mItems.indexOf(item);
            itemView.setTag(itemIndex);
        }

        @Override
        public void onClick(View view) {
            mListener.onItemClick((Integer) view.getTag());
        }
    }

    public interface OnItemClickListener {
        void onItemClick(int index);
    }
}

File: app/src/main/java/org/nitri/opentopo/adapter/WayPointListAdapter.kt
package org.nitri.opentopo.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import org.nitri.opentopo.R
import org.nitri.opentopo.model.WayPointHeaderItem
import org.nitri.opentopo.model.WayPointItem
import org.nitri.opentopo.model.WayPointListItem

class WayPointListAdapter(
    private val mItems: List<WayPointListItem>,
    private val mListener: OnItemClickListener
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    override fun getItemViewType(position: Int): Int {
        return mItems[position].getListItemType()
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, type: Int): RecyclerView.ViewHolder {
        val view: View
        when (type) {
            WayPointListItem.HEADER -> {
                view = LayoutInflater
                    .from(viewGroup.context)
                    .inflate(R.layout.way_point_header_item, viewGroup, false)
                return ViewHolderHeader(view)
            }

            WayPointListItem.WAY_POINT -> {
                view = LayoutInflater
                    .from(viewGroup.context)
                    .inflate(R.layout.way_point_item, viewGroup, false)
                return ViewHolderWayPoint(view)
            }

            else -> return DefaultViewHolder(View(viewGroup.context))  // Should never occur!
        }

    }

    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, pos: Int) {
        val item = mItems[pos]
        (viewHolder as ViewHolder).bindType(item)
    }

    override fun getItemCount(): Int {
        return mItems.size
    }

    abstract class ViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(
        itemView) {
        abstract fun bindType(item: WayPointListItem)
    }

    class ViewHolderHeader internal constructor(itemView: View) : ViewHolder(itemView) {
        private val textView: TextView

        init {
            textView = itemView.findViewById(R.id.textView)
        }

        override fun bindType(item: WayPointListItem) {
            textView.text = (item as WayPointHeaderItem).header
        }
    }

    /**
     * Class for non-null returns
     */
    class DefaultViewHolder(view: View) : ViewHolder(view) {
        override fun bindType(item: WayPointListItem) {
            // NOP
        }
    }

    inner class ViewHolderWayPoint internal constructor(itemView: View) : ViewHolder(itemView),
        View.OnClickListener {
        private val textView: TextView

        init {
            textView = itemView.findViewById(R.id.textView)
            itemView.setOnClickListener(this)
        }

        override fun bindType(item: WayPointListItem) {
            textView.text = (item as WayPointItem).wayPoint.name
            val itemIndex = mItems.indexOf(item)
            itemView.tag = itemIndex
        }

        override fun onClick(view: View) {
            mListener.onItemClick(view.tag as Int)
        }
    }

    interface OnItemClickListener {
        fun onItemClick(index: Int)
    }
}